Add text on principals and agents (#19)
jhoyla/draft-jhoyla-req-mtls-flag: draft-jhoyla-req-mtls-flag.md

# Security Considerations

## Authentication Properties

TLS client authentication is bound to a particular TLS session. This means that
a malicious server cannot manipulate a client in such a way that they can
impersonate said client to any other server. This is due to the matching

> sessions property of TLS.

Reviewer: let's add an informative reference for "matching sessions property of TLS"? https://eprint.iacr.org/2020/1044 comes to mind, but pick your favorite :)

> This differs sharply from e.g. HTTP Signatures, where the receiving server can

Reviewer: Is this the first reference to HTTP signatures? If so, add a reference to @thibmeu's draft here.

relay a signature header to another server. In HTTP Signatures this is

> intentional, as it enables the use of proxies.

Reviewer: ```suggestion
intentional, as it easy to implement in the presence of an HTTP proxy.
```

Author: This changes the meaning. The point is specifically contrasting the fact that it is possible to use HTTP Sigs in the presence of a proxy, but that it is not possible in the case of client auth. The ease of implementation is a red herring.

Reviewer: SGTM. Based on our conversation last Friday (or Thursday?) I thought you wanted to leave open the possibility of request client auth being useful even if there's a proxy. I read this as "if you have an HTTP proxy in the middle, then you're gonna have a bad time", but I suppose that's a misreading too.

Author: So client auth can be useful in the presence of a proxy, but relies on the origin trusting the proxy. This is probably true in the case of a reverse proxy, given that they control your private key, but not necessarily.

This is because they authenticate different things. TLS client authentication
identifies the actor you are talking to directly, whereas HTTP signatures
identifies the actor that triggered some request, but not necessarily the actor
you are in direct contact with.

We describe the difference between these two properties as "principal

> authentication" and "agent authentication". Principals are the initiators of a

Reviewer: Do we have an informative reference for "principal" versus "agent" authentication?:

Author: No, although it would be great to find one. I've found a few other projects that use the word principal, but none that use it in precisely this sense.

Reviewer: Ack, that's a bummer.

particular sequence of actions. It is expected for their authentication
information to be proxied from agent to agent as the relevant actions play out.
For example a crawler might request a webpage from a CDN, and the CDN may relay
the crawler's authentication, in the form of a `Signature` header, to the origin
server. Agents, on the other hand, are the immediate participants in a protocol.

> Each agent in a chain authenticates itself only to its immediate neighbours.

Reviewer: I still think the concept of "immediate neighbors" is a little vague.

The upside of agent authentication is that it is session-matching, i.e. it
cannot be replayed. However that authentication cannot be passed on. The upside
of principal authentication is that it can be passed down an entire chain of

> agents, and each agent can verify it independently. However, it suffers from

Reviewer: "chain of agents": see https://github.com/jhoyla/draft-jhoyla-req-mtls-flag/pull/19#discussion_r2150017380

issues with replays.

In some cases it might be desirable to have both types of authentication, and in
others only one. TLS client auth provides agent authentication and HTTP

> Signatures provide principal authentication. It is possible for both protocols

Reviewer: We need to explain who is the principal and who is the agent ini TLS client auth.

Also, nit: don't shorten authentication to auth
```suggestion
others only one. TLS client authentication provides agent authentication and HTTP
Signatures provide principal authentication. It is possible for both protocols
```

sessions property of TLS {{?DFGS21=DOI.10.1007/s00145-021-09384-1}}.

This differs sharply from e.g. HTTP Signatures
{{?I-D.draft-meunier-web-bot-auth-architecture}}, where the receiving server
can relay a signature header to another server. In HTTP Signatures this is
intentional, as it enables the use of proxies.

This is because they authenticate different things. TLS client authentication
identifies the actor you are talking to directly, whereas HTTP signatures
identifies the actor that triggered some request, but not necessarily the actor
you are in direct contact with.

We describe the difference between these two properties as "principal
authentication" and "agent authentication". Principals are the initiators of a
particular sequence of actions. It is expected for their authentication
information to be proxied from agent to agent as the relevant actions play out.
For example a crawler might request a webpage from a CDN, and the CDN may relay
the crawler's authentication, in the form of a `Signature` header, to the
origin server. Agents, on the other hand, are the immediate participants in a
protocol. Each agent in a chain authenticates itself only to its direct peers,

> i.e. those with whom it establishes a TLS connection.

Reviewer 2: i really like this section an framing
if you're willing, it would make for a great addition to the glossary and the related consideration discussed https://github.com/thibmeu/draft-meunier-glossary-somehow/issues/5

The upside of agent authentication is that it is session-matching, i.e. it
cannot be replayed. However that authentication cannot be passed on. The upside
of principal authentication is that a single authority (the principal) can
create a signature, and even if there is a long chain of proxies, each acting
as an independent agent, the signature can be passed from one agent to another,
and each agent can verify it independently. However, it suffers from issues

> with replays, see Section {{http-sigs}}.

Reviewer: This would render as "Section Section XXX".
```suggestion
with replays, see {{http-sigs}}.
```

In some cases it might be desirable to have both types of authentication, and
in others only one. TLS client authentication provides agent authentication,
the client leaf certificate acting to authenticate the agent, and HTTP
Signatures provide principal authentication, the signature authenticating the
principal. It is possible for both protocols to be set up to provide the other
as well, although this requires special deployment considerations. For example,
the TLS Client can be issued with a certificate chain whose leaf provides
authentication of the client, and whose CA provides authentication of the
principal. Similarly HTTP Signatures can provide agent authentication by

> including a binding from the exporter key interface in the signature.

Reviewer 2: probably worth mentioning [RFC 8473 Token Binding over HTTP](https://datatracker.ietf.org/doc/html/rfc8473)

that chain to the bots trust anchors and those that chain to other sets of
trust anchors are disjoint.

> ## Principals vs Agents

Reviewer: Not a great title IMO, as "principal" and "agent" aren't' defined yet.

I actually had something else in mind for this text. Tell me what you think about this. I think we have two main goals, the first to describe the intended security properties of request client auth, and the second to point out the shortcomings of HTTP signatures. (The answer of course is that HTTP sigs authenticates the principal, not the agent!)

Therefore I think this text should be the _start_ of the security considerations section. The current text, which gets into considerations for deploying the mechanism, could go in its own subsection. WDYT?

```suggestion
## Principals vs Agents
```

TLS client authentication is bound to a particular TLS session. This means that
a malicious server cannot manipulate a client in such a way that they can
impersonate said client to any other server. This is due to the matching
sessions property of TLS. This, however, does have some downsides. A server
behind a reverse proxy cannot prove to the origin that it saw and appropriately
verified a particular client.

This differs sharply from e.g. HTTP Signatures, where tokens can be passed from
actor to actor, and can always be verified. This is because they authenticate

> different things.

Reviewer: Give an example of a concrete attack here, e.g. cut-and-paste the multi-CDN example from the alternative-appraoches section.

We describe the difference between these two properties as "principal
authentication" and "agent authentication". Principals are the initiators of a
particular sequence of actions. It is expected for their authentication

> information to be passed from agent to agent as the relevant actions play out.

Reviewer: "from agent to agent": What do you mean here? Can you give an example?

Agents, on the other hand, are the immediate participants in a protocol. Each
agent in a chain authenticates itself only to its immediate neighbours.

The upside of agent authentication is that it is session-matching, i.e. it

> cannot be replayed. However it also cannot be passed on. The upside of principal

Reviewer: Why would someone want to pass "it" on? Is this the reverse/forward proxy case?

bit: Also, what is "it"? "authentication of the agent"?

> authentication is that it can be passed through an entire chain of actors, and

Reviewer: "entire chain of actors": for example? I'm wary about introducing too many hypotheticals.

nit: If by "actor" you mean "agent", then let's say "agent" to avoid using too many words for the same thing.

that chain to the bots trust anchors and those that chain to other sets of
trust anchors are disjoint.

> 

Reviewer: nit: extraneous change
```suggestion
```
